File: sdk/python/src/spacerouter/payment/consumer_settlement.py
"""Consumer-side Leg 1 settlement client.

Replaces the broken in-band TCP receipt exchange. After a SPACE-paid
proxy request the Gateway stashes an unsigned Leg 1 receipt in
``pending_client_receipts``. This client pulls the pending receipts
via ``GET /leg1/pending``, signs each with EIP-712, and submits back
via ``POST /leg1/sign`` — the same out-of-band pattern Leg 2 uses
between the Provider and the Coord API.

The client is stateless: every call authenticates with a fresh EIP-191
signature over ``space-router:leg1-<verb>:<addr>:<ts>``. Safe to call
repeatedly; the Gateway side is idempotent.

Usage:
    settler = ConsumerSettlementClient(
        gateway_url="https://gateway:8081",
        private_key="0x...",
    )
    result = await settler.sync_receipts()
    # {"accepted": [uuid, ...], "rejected": [{"request_uuid", "reason"}]}
"""

from __future__ import annotations

import logging
import time
from typing import Any

import httpx
import tenacity
from eth_account import Account
from eth_account.messages import encode_defunct

from spacerouter.payment.eip712 import EIP712Domain, Receipt, sign_receipt

logger = logging.getLogger(__name__)


class _RetryableHTTPError(Exception):
    """Wraps a 5xx response so tenacity can retry on it.

    Tenacity's ``retry_if_exception_type`` operates on raised exceptions,
    not on returned values. We raise this internally between attempts and
    convert back to ``httpx.HTTPStatusError`` after the final failure so
    callers see the standard exception type with the verbatim body.
    """

    def __init__(self, response: httpx.Response) -> None:
        self.response = response
        body = response.text
        super().__init__(
            f"HTTP {response.status_code} from "
            f"{response.request.method} {response.request.url}: {body}"
        )


# Per spec §9: bounded retry, default 3 attempts, 200ms / 1s / 5s.
# We use exponential backoff seeded so the first wait is 200ms, then 1s, 5s.
_RETRY_KW = dict(
    stop=tenacity.stop_after_attempt(3),
    wait=tenacity.wait_exponential(multiplier=0.2, min=0.2, max=5.0),
    retry=tenacity.retry_if_exception_type(
        (httpx.TransportError, _RetryableHTTPError),
    ),
    reraise=True,
)


class ConsumerSettlementClient:
    """Fetches pending Leg 1 receipts, signs, submits."""

    def __init__(
        self,
        gateway_url: str,
        private_key: str,
        timeout: float = 10.0,
        verify: bool = True,
    ) -> None:
        """
        Parameters
        ----------
        gateway_url : str
            Gateway management URL (e.g. ``https://gateway:8081``).
        private_key : str
            Consumer's wallet private key. Signs both EIP-191 auth and
            EIP-712 receipts.
        timeout : float
            HTTP timeout per call.
        verify : bool
            TLS verification (set False only for local dev).
        """
        self._gateway_url = gateway_url.rstrip("/")
        self._account = Account.from_key(private_key)
        self._private_key = private_key
        self._timeout = timeout
        self._verify = verify

    @property
    def address(self) -> str:
        return self._account.address

    def _auth_sig(self, verb: str, ts: int) -> str:
        """EIP-191 sig over ``space-router:leg1-<verb>:<addr>:<ts>``."""
        msg = f"space-router:leg1-{verb}:{self._account.address.lower()}:{ts}"
        signed = self._account.sign_message(encode_defunct(text=msg))
        return "0x" + signed.signature.hex()

    async def fetch_pending(self, limit: int = 50) -> dict[str, Any]:
        """Pull unsigned Leg 1 receipts owed by this consumer.

        Returns a dict with ``receipts`` (list) and ``domain`` (EIP-712
        domain to sign under).

        Retries on transport errors and HTTP 5xx with exponential backoff
        (3 attempts, 200ms / 1s / 5s) per spec §9. The verbatim response
        body is surfaced in the raised error message on final failure.
        """
        @tenacity.retry(**_RETRY_KW)
        async def _do() -> dict[str, Any]:
            ts = int(time.time())
            params = {
                "address": self._account.address,
                "ts": ts,
                "sig": self._auth_sig("list-pending", ts),
                "limit": limit,
            }
            async with httpx.AsyncClient(
                timeout=self._timeout, verify=self._verify,
            ) as client:
                r = await client.get(
                    f"{self._gateway_url}/leg1/pending", params=params,
                )
            if 500 <= r.status_code < 600:
                raise _RetryableHTTPError(r)
            r.raise_for_status()
            return r.json()

        try:
            return await _do()
        except _RetryableHTTPError as exc:
            # Surface body verbatim via the standard httpx exception.
            exc.response.raise_for_status()
            raise  # unreachable

    async def submit_signatures(
        self, signatures: list[dict[str, str]],
    ) -> dict[str, Any]:
        """POST ``{request_uuid, signature}`` pairs back to the gateway.

        Returns ``{accepted: [uuid, ...], rejected: [{request_uuid, reason}]}``.

        Retries on transport errors and HTTP 5xx with exponential backoff
        (3 attempts, 200ms / 1s / 5s) per spec §9. The verbatim response
        body is surfaced in the raised error message on final failure.
        """
        if not signatures:
            return {"accepted": [], "rejected": []}

        @tenacity.retry(**_RETRY_KW)
        async def _do() -> dict[str, Any]:
            ts = int(time.time())
            body = {
                "address": self._account.address,
                "ts": ts,
                "sig": self._auth_sig("sign", ts),
                "signatures": signatures,
            }
            async with httpx.AsyncClient(
                timeout=self._timeout, verify=self._verify,
            ) as client:
                r = await client.post(
                    f"{self._gateway_url}/leg1/sign", json=body,
                )
            if 500 <= r.status_code < 600:
                raise _RetryableHTTPError(r)
            r.raise_for_status()
            return r.json()

        try:
            return await _do()
        except _RetryableHTTPError as exc:
            exc.response.raise_for_status()
            raise  # unreachable

    async def sync_receipts(self, limit: int = 50) -> dict[str, Any]:
        """One-shot: fetch pending, sign each, submit, return the outcome.

        Safe to call anytime — after each proxy request for real-time
        settlement, or periodically for batch settlement of accumulated
        receipts.
        """
        pending = await self.fetch_pending(limit=limit)
        receipts = pending.get("receipts", [])
        if not receipts:
            return {"accepted": [], "rejected": [], "pending_count": 0}

        domain_dict = pending["domain"]
        domain = EIP712Domain(
            name=domain_dict["name"],
            version=domain_dict["version"],
            chain_id=int(domain_dict["chainId"]),
            verifying_contract=domain_dict["verifyingContract"],
        )

        signatures: list[dict[str, str]] = []
        for row in receipts:
            receipt = Receipt(
                client_address=row["client_address"],
                node_address=row["node_address"],
                request_uuid=row["request_uuid"],
                data_amount=int(row["data_amount"]),
                total_price=int(row["total_price"]),
            )
            try:
                sig = sign_receipt(self._private_key, receipt, domain)
            except Exception as exc:
                logger.exception(
                    "Failed to sign Leg 1 receipt uuid=%s: %s",
                    receipt.request_uuid, exc,
                )
                continue
            signatures.append({
                "request_uuid": receipt.request_uuid,
                "signature": sig,
            })

        result = await self.submit_signatures(signatures)
        result["pending_count"] = len(receipts)
        logger.info(
            "Leg 1 sync: %d pending → %d accepted, %d rejected",
            len(receipts),
            len(result.get("accepted", [])),
            len(result.get("rejected", [])),
        )
        return result

File: sdk/js/src/payment/consumerSettlement.ts
/**
 * Consumer-side Leg 1 settlement broker client.
 *
 * Implements the out-of-band Leg 1 receipt-exchange protocol described in
 * `internal-docs/v1.5-consumer-protocol.md` §3-4 and §9 (negative paths):
 *
 *   GET  {gateway_mgmt}/leg1/pending  ?address&ts&sig[&limit]
 *   POST {gateway_mgmt}/leg1/sign     {address, ts, sig, signatures: [...]}
 *
 * EIP-191 auth messages (§4):
 *   space-router:leg1-list-pending:<addr-lowercase>:<ts>
 *   space-router:leg1-sign:<addr-lowercase>:<ts>
 *
 * Strict-mode rejection raises `SettlementRejectedError`; non-strict
 * surfaces the full reasons list so callers can decide.
 */

import { type ClientPaymentWallet } from "./clientWallet.js";
import { type EIP712Domain, type Receipt } from "./eip712.js";
import { SettlementRejectedError } from "../errors.js";
import { fetch as undiciFetch } from "undici";

// ---------------------------------------------------------------------------
// Wire types — snake_case per protocol §8.
// ---------------------------------------------------------------------------

interface WireDomain {
  name: string;
  version: string;
  chainId: number;
  verifyingContract: `0x${string}`;
}

interface WireReceipt {
  request_uuid: string;
  client_address: string;
  node_address: string;
  data_amount: number | string;
  total_price: number | string;
  tunnel_request_id?: string;
  created_at?: string;
}

interface PendingResponse {
  receipts: WireReceipt[];
  domain: WireDomain;
}

interface WireSignatureEntry {
  request_uuid: string;
  signature: `0x${string}`;
}

interface WireSignResponse {
  accepted: string[];
  rejected: Array<{ request_uuid: string; reason: string }>;
}

// ---------------------------------------------------------------------------
// Public types
// ---------------------------------------------------------------------------

export interface ConsumerSettlementOptions {
  gatewayMgmtUrl: string;
  // TODO(track-c): swap stub for real ClientPaymentWallet from Track C.
  wallet: ClientPaymentWallet;
}

export interface PendingFetchResult {
  receipts: Receipt[];
  domain: EIP712Domain;
}

export interface SignatureSubmission {
  requestUuid: string;
  signature: `0x${string}`;
}

export interface SignatureSubmitResult {
  accepted: string[];
  rejected: Array<{ requestUuid: string; reason: string }>;
}

export interface SubmitOptions {
  strict?: boolean;
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/** Parse a wire `uint256` (number or decimal-string) as bigint per §8. */
function parseUint256(v: number | string): bigint {
  if (typeof v === "number") {
    if (!Number.isInteger(v) || v < 0) {
      throw new Error(`expected non-negative integer uint256, got ${v}`);
    }
    return BigInt(v);
  }
  return BigInt(v);
}

function decodeReceipt(raw: WireReceipt): Receipt {
  return {
    clientAddress: raw.client_address.toLowerCase() as `0x${string}`,
    nodeAddress: raw.node_address.toLowerCase() as `0x${string}`,
    requestUUID: raw.request_uuid,
    dataAmount: parseUint256(raw.data_amount),
    totalPrice: parseUint256(raw.total_price),
  };
}

function decodeDomain(raw: WireDomain): EIP712Domain {
  return {
    name: raw.name,
    version: raw.version,
    // Wire format is JSON number; Track C's eip712 domain typedata expects
    // bigint per viem's signTypedData signature.
    chainId: BigInt(raw.chainId),
    verifyingContract: raw.verifyingContract,
  };
}

function nowSeconds(): number {
  return Math.floor(Date.now() / 1000);
}

// ---------------------------------------------------------------------------
// Bounded retry — protocol §9: 3 attempts, 200ms / 1s / 5s, retry on
// network errors and HTTP 5xx; surface body verbatim on final failure.
// ---------------------------------------------------------------------------

const RETRY_DELAYS_MS = [200, 1000, 5000] as const;

async function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

interface FetchAttemptResult {
  status: number;
  body: string;
}

async function fetchWithRetry(
  url: string,
  init: { method: string; headers?: Record<string, string>; body?: string },
): Promise<FetchAttemptResult> {
  let lastError: unknown;
  let lastResult: FetchAttemptResult | undefined;

  for (let attempt = 0; attempt < RETRY_DELAYS_MS.length; attempt += 1) {
    if (attempt > 0) {
      await sleep(RETRY_DELAYS_MS[attempt - 1]!);
    }

    try {
      const resp = await undiciFetch(url, {
        method: init.method,
        headers: init.headers,
        body: init.body,
      });
      const text = await resp.text();
      const result: FetchAttemptResult = { status: resp.status, body: text };

      if (resp.status >= 500 && resp.status < 600) {
        lastResult = result;
        continue;
      }
      return result;
    } catch (err) {
      // Network-level failure (TypeError per fetch spec, undici may also throw).
      if (err instanceof TypeError) {
        lastError = err;
        continue;
      }
      throw err;
    }
  }

  if (lastResult) return lastResult;
  throw lastError ?? new Error("fetchWithRetry exhausted with no result");
}

// ---------------------------------------------------------------------------
// ConsumerSettlementClient
// ---------------------------------------------------------------------------

export class ConsumerSettlementClient {
  private readonly _gatewayMgmtUrl: string;
  // TODO(track-c): swap stub for real ClientPaymentWallet from Track C.
  private readonly _wallet: ClientPaymentWallet;

  constructor(options: ConsumerSettlementOptions) {
    this._gatewayMgmtUrl = options.gatewayMgmtUrl.replace(/\/+$/, "");
    this._wallet = options.wallet;
  }

  // -------------------------------------------------------------------------
  // GET /leg1/pending
  // -------------------------------------------------------------------------

  async fetchPending(limit = 50): Promise<PendingFetchResult> {
    const addr = this._wallet.address.toLowerCase();
    const ts = nowSeconds();
    const message = `space-router:leg1-list-pending:${addr}:${ts}`;
    const sig = await this._wallet.signMessage(message);

    const qs = new URLSearchParams({
      address: addr,
      ts: String(ts),
      sig,
      limit: String(limit),
    });
    const url = `${this._gatewayMgmtUrl}/leg1/pending?${qs.toString()}`;

    const { status, body } = await fetchWithRetry(url, { method: "GET" });
    if (status < 200 || status >= 300) {
      throw new Error(
        `GET /leg1/pending failed: HTTP ${status} — ${body}`,
      );
    }

    const parsed = JSON.parse(body) as PendingResponse;
    return {
      receipts: (parsed.receipts ?? []).map(decodeReceipt),
      domain: decodeDomain(parsed.domain),
    };
  }

  // -------------------------------------------------------------------------
  // POST /leg1/sign
  // -------------------------------------------------------------------------

  async submitSignatures(
    sigs: SignatureSubmission[],
    opts?: SubmitOptions,
  ): Promise<SignatureSubmitResult> {
    const addr = this._wallet.address.toLowerCase();
    const ts = nowSeconds();
    const message = `space-router:leg1-sign:${addr}:${ts}`;
    const authSig = await this._wallet.signMessage(message);

    const wireSigs: WireSignatureEntry[] = sigs.map((s) => ({
      request_uuid: s.requestUuid,
      signature: s.signature,
    }));

    const body = JSON.stringify({
      address: addr,
      ts,
      sig: authSig,
      signatures: wireSigs,
    });

    const url = `${this._gatewayMgmtUrl}/leg1/sign`;
    const { status, body: respBody } = await fetchWithRetry(url, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body,
    });

    if (status < 200 || status >= 300) {
      throw new Error(
        `POST /leg1/sign failed: HTTP ${status} — ${respBody}`,
      );
    }

    const parsed = JSON.parse(respBody) as WireSignResponse;
    const result: SignatureSubmitResult = {
      accepted: parsed.accepted ?? [],
      rejected: (parsed.rejected ?? []).map((r) => ({
        requestUuid: r.request_uuid,
        reason: r.reason,
      })),
    };

    if (opts?.strict && result.rejected.length > 0) {
      throw new SettlementRejectedError(result.rejected);
    }
    return result;
  }

  // -------------------------------------------------------------------------
  // syncReceipts: fetch + sign + submit (one-shot)
  // -------------------------------------------------------------------------

  async syncReceipts(
    limit = 50,
    opts?: SubmitOptions,
  ): Promise<SignatureSubmitResult> {
    const { receipts, domain } = await this.fetchPending(limit);
    if (receipts.length === 0) {
      return { accepted: [], rejected: [] };
    }

    const sigs: SignatureSubmission[] = [];
    for (const receipt of receipts) {
      const signature = await this._wallet.signReceipt(receipt, domain);
      sigs.push({ requestUuid: receipt.requestUUID, signature });
    }
    return this.submitSignatures(sigs, opts);
  }
}
